### PR TITLE
#3 Add XML formatting for record key and value

### PR DESCRIPTION
**Pull request by tbrv**
Added checkbox to format XML for key and value

### CODE REVIEW

File: src/main/java/kafkavisualizer/details/consumer/ConsumerEventPane.java [hunk 1]
```diff
@@ -39,9 +44,11 @@ public ConsumerEventPane() {
         var keyToolbar = new JToolBar();
         keyTextAreaWordWrapCheckBox = new JCheckBox("Word Wrap");
         keyTextAreaFormatJSONCheckBox = new JCheckBox("Format JSON");
+        keyTextAreaFormatXMLCheckBox = new JCheckBox("Format JSON");
```
Comment (winne42) [inline src/main/java/kafkavisualizer/details/consumer/ConsumerEventPane.java:47]: I think you mean "Format XML" here?

File: src/main/java/kafkavisualizer/details/consumer/ConsumerDetailsController.java [hunk 2]
```diff
@@ -148,28 +196,38 @@ public void updateConsumerPane() {
...
     }
 
+    private String formatString(Format format, String str) {
+        if (format == Format.XML) {
+            if (xmlPrettyFormatter == null) {
+                return str;
+            } else {
+                return xmlPrettyFormatter.format(str);
+            }
+        } else if (valueFormat == Format.JSON) {
```
Comment (winne42) [inline src/main/java/kafkavisualizer/details/consumer/ConsumerDetailsController.java:225]: I think you mean `format` here, not `valueFormat`

**Comment by kumait** [comment]
This implementation has some problems

- XML header is added to the message when formatting it.
- The UI uses check boxes not the way they are intended to be used.
- No need for a separate class to do the format, adding a method to Utils is more consistent.

File: src/main/java/kafkavisualizer/XmlPrettyFormatter.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    private Transformer createTransformer() {
+        try {
+            var factory = TransformerFactory.newInstance();
+            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
+
+            var xmlTransformer = factory.newTransformer();
+            // prohibit the use of all protocols by external entities:
```
Comment (winne42) [inline src/main/java/kafkavisualizer/XmlPrettyFormatter.java:53]: comment does not match code below

File: src/main/java/kafkavisualizer/XmlPrettyFormatter.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        try {
+            Document doc = documentBuilder.parse(new InputSource(new StringReader(xml)));
+
+            DOMSource source = new DOMSource(doc);
+            StreamResult result = new StreamResult(new StringWriter());
+            transformer.transform(source, result);
+            return result.getWriter().toString();
+        } catch (SAXException | IOException | TransformerException e) {
+            System.err.println("Error prettifying xml: " + e.getMessage());
```
Comment (winne42) [inline src/main/java/kafkavisualizer/XmlPrettyFormatter.java:76]: pretty-printing? (there is **no** way, to make XML pretty :-D)

**Comment by tbrv** [comment]
@kumait 

> This implementation has some problems
> * XML header is added to the message when formatting it.

I don't understand this. Could you clarify? 

> * The UI uses check boxes not the way they are intended to be used.

What would you prefer? Radio-buttons or a dropdown menu? 





> * No need for a separate class to do the format, adding a method to Utils is more consistent.

I could move this to Utils but I would strongly recomment to have the formatter separate class, as it is not trivial as the JSON implementation is.

**Comment by winne42** [comment]
I assume @kumait means the XML declaration? Adding it could be avoided using
```java
xmlTransformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
```

**Pull request closed without merging**